File: 350.intersection-of-two-arrays-ii.cpp
class Solution {
public:
    /* Approach: Sorting + Two Pointer, Complexity: O(nlogn + mlogm), O(1) */
    /* Tip: Sort both arrays. Use two pointers traversing the two arrays at the
     *      same time. Move the pointer which points at a smaller element and 
     *      then move. Pointer to smaller element is moved as it is still 
     *      possible to find the larger element in other array but it is not 
     *      possible to find the smaller element to the right of a larger
     *      element as the arrays are sorted. On finding a common element,
     *      move both pointers to find other common elements.
     */

    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());

        vector<int> common;
        int i, j;
        i = j = 0;

        while (i < nums1.size() && j < nums2.size()) {
            if (nums1[i] < nums2[j])
                ++i;
            else if (nums1[i] > nums2[j])
                ++j;
            else {
                common.push_back(nums1[i]);
                ++i;
                ++j;
            }
        }
        return common;
    }
};

File: 350.intersection-of-two-arrays-ii.py
class Solution:
    # Approach: Hashmap, Complexity: O(m+n), O(min(m, n))
    # Tip: Record the elements of the smaller array along with their frequencies
    #      in a hashmap. Iterate over the other array. If a match is found in
    #      the hashmap, decrement its count. This is because one instance of the
    #      element in the smaller array has been exhausted i.e. an element common
    #      with it has been found in other array.

    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        if len(nums1) < len(nums2):
            nums_small, nums_big = nums1, nums2
        else:
            nums_small, nums_big = nums2, nums1

        nums_small_freq = {}
        for num in nums_small:
            nums_small_freq[num] = nums_small_freq.get(num, 0) + 1

        common = []
        nums_small = set(nums_small)

        for num in nums_big:
            if num in nums_small_freq:
                common.append(num)
                if nums_small_freq[num] == 1:
                    nums_small_freq.pop(num)
                else:
                    nums_small_freq[num] -= 1
        return common
